rfcs/RFC-project-rules.md

```diff
@@ -0,0 +1,75 @@
+# Enforcing contribution rules for repository stability with a merging bot
+
+**Authors:**
+
+* Red Hat: @leseb
+
+## Summary
+
+To ensure the stability, security, and maintainability of the project, we propose enforcing the following contribution rules. These guidelines help maintain code quality, prevent accidental regressions, and improve collaboration across contributors.
+
+We recognize that many of these rules may already be considered common sense by experienced contributors. However, making them explicit and automating their enforcement ensures consistency across all contributors, avoids misunderstandings, and streamlines the development process. The goal is not to restrict contributions but to create a smoother workflow that benefits everyone.
+
+Rather than relying solely on manual enforcement, we propose using a merging bot to handle these rules automatically, reducing the risk of human error and keeping the process efficient.
+
+## Proposed Rules & Automated Enforcement
+
+1. Do not push directly to main
+[Branch protection](https://docs.github.com/en/repositories/configuring-branches-and-merges-in-your-repository/managing-protected-branches/managing-a-branch-protection-rule) settings will prevent direct pushes to main, requiring all changes to go through pull requests (PRs).
+
+2. Do not self-merge
+The merging bot will ensure that a PR has at least one approval from another maintainer before merging.
+
+3. Do not merge if CI is failing
+The bot will block merging if the CI checks do not pass, ensuring that only validated code is integrated. Maintainers with write access can override the bot and merge the PR anyway if needed.
+
+4. Do not use main for development-use a fork instead
+Documentation will guide contributors on forking best practices, and a workflow can optionally reject PRs not coming from a fork if required.
+
+5. Require code to be up to date with main before merging
+The bot will check if the PR is outdated compared to main and require a rebase or merge with the latest main branch before allowing the merge.
+
+By automating these rules through a merging bot, we ensure a consistent, high-quality, and efficient development process with minimal manual oversight.
+
+## Sample bot configuration
+
+[Mergify](https://mergify.com/) is a merging bot that can be configured to enforce the proposed rules.
+
+Here a sample configuration::
+
+```yaml
+pull_request_rules:
+- name: auto-merge
+  conditions:
+    - "#approved-reviews-by>=1"
+    - check-success=pre-commit
+    - check-success="Facebook CLA Check"
+    - check-success="unit-tests"
+    - check-success="integration-tests"
+
+  actions:
+    merge:
+      method: rebase # rebase and merge
```

Comment:
AFAIU this will also change how patch series are merged. (Leaving separate patches.) Correct?

Which is I'm all for it, but it's a change for the project (where I think currently we always squash + use PR description as the final commit message.) I think this change should be called out here, so that it's not news for others. And we may have to update contributors' guide to reflect the need to maintain healthy git habits for patch series and patch commit messages.

Reply:
Yes, that's why I'm highlighting 2 solutions. To be less disruptive, we could go with `method: squash` to mimic the current behavior.